### PR TITLE
Generalize QR solver to matrices

### PR DESCRIPTION
**Pull request by peddie**
whaddaya think

### CODE REVIEW

**Comment by kovach** [comment]
I think I fixed the build issue (which I introduced) could you rebase onto master?

File: examples/qr.plv [hunk 1]
```diff
@@ -56,26 +56,27 @@ qr_update {m, n}
         A[i-2 .. i-1, k-1] <- rot^T * v;
       );
 
-      -- Rotate b vector
-      v := b[i-2 .. i-1];
-      b[i-2 .. i-1] <- rot^T * v;
-
+      -- Rotate b vector TODO(MP)
```
Comment (scottswift) [inline examples/qr.plv]: please delete TODO

File: examples/qr.plv [hunk 2]
```diff
@@ -17,31 +17,31 @@ static givens (a :: double) (b :: double) :: double[2,2] := (
...
 
   :: s8 := (
 
   qr_update (inout b) (inout A);
 
   -- A is now upper triangular; backsolve it into b
-  code := backsolve A[0:n, 0:n] (inout b[0:n]);
+  code := backsolve A[0:n, 0:n] (inout b[0:n, :]);
 
   -- Solution stored in first n elements
```
Comment (scottswift) [inline examples/qr.plv]: maybe change 'elements' to 'rows' here and below

File: examples/qr_test.plv [hunk 3]
```diff
@@ -9,22 +9,44 @@ main () :: s32 := (
...
   v := b;
 
-  residual := -1.0;
+  residual :: double[2];
+
+  solution :: double[3, 2];
+
+  ret := qr_solve (inout m) (inout b) (out solution) (out residual);
+
+  -- TODO(MP): check the actual solution, don't just print it
```
Comment (scottswift) [inline examples/qr_test.plv]: delete todo or add assert for good measure

**Pull request merged**